### PR TITLE
Implement Finetuning from MRI data

### PR DESCRIPTION
**Pull request by m-krastev**
Tested on LoRA model.

### CODE REVIEW

File: adapt_med_seg/models/segvol_lora.py [hunk 1]
```diff
@@ -31,7 +31,8 @@ def __init__(self, config: SegVolConfig, r=8, alpha = 8, dropout = 0.0):
             lora_dropout=dropout,
```
Comment (MJHamar) [inline adapt_med_seg/models/segvol_lora.py:31]: Note that selecting `q_proj` and `v_proj` will only train the mask decoder

Comment (MJHamar) [reply adapt_med_seg/models/segvol_lora.py:31]: Here is the current model:
```
SegVolLORA(
  (model): PeftModel(
    (base_model): LoraModel(
      (model): SegVol(
        (image_encoder): ViT(
          (patch_embedding): PatchEmbeddingBlock(
            (patch_embeddings): Sequential(
              (0): Rearrange('b c (h p1) (w p2) (d p3) -> b (h w d) (p1 p2 p3 c)', p1=4, p2=16, p3=16)
              (1): Linear(in_features=1024, out_features=768, bias=True)
            )
            (dropout): Dropout(p=0.0, inplace=False)
          )
          (blocks): ModuleList(
            (0-11): 12 x TransformerBlock(
              (mlp): MLPBlock(
                (linear1): Linear(in_features=768, out_features=3072, bias=True)
                (linear2): Linear(in_features=3072, out_features=768, bias=True)
                (fn): GELU(approximate='none')
                (drop1): Dropout(p=0.0, inplace=False)
                (drop2): Dropout(p=0.0, inplace=False)
              )
              (norm1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
              (attn): SABlock(
                (out_proj): Linear(in_features=768, out_features=768, bias=True)
                (qkv): Linear(in_features=768, out_features=2304, bias=False)
                (input_rearrange): Rearrange('b h (qkv l d) -> qkv b l h d', qkv=3, l=12)
                (out_rearrange): Rearrange('b h l d -> b l (h d)')
                (drop_output): Dropout(p=0.0, inplace=False)
                (drop_weights): Dropout(p=0.0, inplace=False)
              )
              (norm2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
            )
          )
          (norm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        )
        (mask_decoder): MaskDecoder(
          (transformer): TwoWayTransformer(
            (layers): ModuleList(
              (0-1): 2 x TwoWayAttentionBlock(
                (self_attn): Attention(
                  (q_proj): lora.Linear(
                    (base_layer): Linear(in_features=768, out_features=768, bias=True)
                    (lora_dropout): ModuleDict(
                      (default): Identity()
                    )
                    (lora_A): ModuleDict(
                      (default): Linear(in_features=768, out_features=8, bias=False)
                    )
                    (lora_B): ModuleDict(
                      (default): Linear(in_features=8, out_features=768, bias=False)
                    )
                    (lora_embedding_A): ParameterDict()
                    (lora_embedding_B): ParameterDict()
                  )
                  (k_proj): Linear(in_features=768, out_features=768, bias=True)
                  (v_proj): lora.Linear(
                    (base_layer): Linear(in_features=768, out_features=768, bias=True)
                    (lora_dropout): ModuleDict(
                      (default): Identity()
                    )
                    (lora_A): ModuleDict(
                      (default): Linear(in_features=768, out_features=8, bias=False)
                    )
                    (lora_B): ModuleDict(
                      (default): Linear(in_features=8, out_features=768, bias=False)
                    )
                    (lora_embedding_A): ParameterDict()
                    (lora_embedding_B): ParameterDict()
                  )
                  (out_proj): Linear(in_features=768, out_features=768, bias=True)
                )
                (norm1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
                (cross_attn_token_to_image): Attention(
                  (q_proj): lora.Linear(
                    (base_layer): Linear(in_features=768, out_features=384, bias=True)
                    (lora_dropout): ModuleDict(
                      (default): Identity()
                    )
                    (lora_A): ModuleDict(
                      (default): Linear(in_features=768, out_features=8, bias=False)
                    )
                    (lora_B): ModuleDict(
                      (default): Linear(in_features=8, out_features=384, bias=False)
                    )
                    (lora_embedding_A): ParameterDict()
                    (lora_embedding_B): ParameterDict()
                  )
                  (k_proj): Linear(in_features=768, out_features=384, bias=True)
                  (v_proj): lora.Linear(
                    (base_layer): Linear(in_features=768, out_features=384, bias=True)
                    (lora_dropout): ModuleDict(
                      (default): Identity()
                    )
                    (lora_A): ModuleDict(
                      (default): Linear(in_features=768, out_features=8, bias=False)
                    )
                    (lora_B): ModuleDict(
                      (default): Linear(in_features=8, out_features=384, bias=False)
                    )
                    (lora_embedding_A): ParameterDict()
                    (lora_embedding_B): ParameterDict()
                  )
                  (out_proj): Linear(in_features=384, out_features=768, bias=True)
                )
                (norm2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
                (mlp): MLPBlock(
                  (lin1): Linear(in_features=768, out_features=2048, bias=True)
                  (lin2): Linear(in_features=2048, out_features=768, bias=True)
                  (act): ReLU()
                )
                (norm3): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
                (norm4): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
                (cross_attn_image_to_token): Attention(
                  (q_proj): lora.Linear(
                    (base_layer): Linear(in_features=768, out_features=384, bias=True)
                    (lora_dropout): ModuleDict(
                      (default): Identity()
                    )
                    (lora_A): ModuleDict(
                      (default): Linear(in_features=768, out_features=8, bias=False)
                    )
                    (lora_B): ModuleDict(
                      (default): Linear(in_features=8, out_features=384, bias=False)
                    )
                    (lora_embedding_A): ParameterDict()
                    (lora_embedding_B): ParameterDict()
                  )
                  (k_proj): Linear(in_features=768, out_features=384, bias=True)
                  (v_proj): lora.Linear(
                    (base_layer): Linear(in_features=768, out_features=384, bias=True)
                    (lora_dropout): ModuleDict(
                      (default): Identity()
                    )
                    (lora_A): ModuleDict(
                      (default): Linear(in_features=768, out_features=8, bias=False)
                    )
                    (lora_B): ModuleDict(
                      (default): Linear(in_features=8, out_features=384, bias=False)
                    )
                    (lora_embedding_A): ParameterDict()
                    (lora_embedding_B): ParameterDict()
                  )
                  (out_proj): Linear(in_features=384, out_features=768, bias=True)
                )
              )
            )
            (final_attn_token_to_image): Attention(
              (q_proj): lora.Linear(
                (base_layer): Linear(in_features=768, out_features=384, bias=True)
                (lora_dropout): ModuleDict(
                  (default): Identity()
                )
                (lora_A): ModuleDict(
                  (default): Linear(in_features=768, out_features=8, bias=False)
                )
                (lora_B): ModuleDict(
                  (default): Linear(in_features=8, out_features=384, bias=False)
                )
                (lora_embedding_A): ParameterDict()
                (lora_embedding_B): ParameterDict()
              )
              (k_proj): Linear(in_features=768, out_features=384, bias=True)
              (v_proj): lora.Linear(
                (base_layer): Linear(in_features=768, out_features=384, bias=True)
                (lora_dropout): ModuleDict(
                  (default): Identity()
                )
                (lora_A): ModuleDict(
                  (default): Linear(in_features=768, out_features=8, bias=False)
                )
                (lora_B): ModuleDict(
                  (default): Linear(in_features=8, out_features=384, bias=False)
                )
                (lora_embedding_A): ParameterDict()
                (lora_embedding_B): ParameterDict()
              )
              (out_proj): Linear(in_features=384, out_features=768, bias=True)
            )
            (norm_final_attn): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
          )
          (iou_token): Embedding(1, 768)
          (mask_tokens): Embedding(4, 768)
          (output_upscaling): Sequential(
            (0): ConvTranspose3d(768, 192, kernel_size=(2, 2, 2), stride=(2, 2, 2))
            (1): LayerNorm((192, 16, 32, 32), eps=1e-05, elementwise_affine=True)
            (2): GELU(approximate='none')
            (3): ConvTranspose3d(192, 96, kernel_size=(2, 2, 2), stride=(2, 2, 2))
            (4): GELU(approximate='none')
          )
          (output_hypernetworks_mlps): ModuleList(
            (0-3): 4 x MLP(
              (layers): ModuleList(
                (0-1): 2 x Linear(in_features=768, out_features=768, bias=True)
                (2): Linear(in_features=768, out_features=96, bias=True)
              )
            )
          )
          (iou_prediction_head): MLP(
            (layers): ModuleList(
              (0): Linear(in_features=768, out_features=256, bias=True)
              (1): Linear(in_features=256, out_features=256, bias=True)
              (2): Linear(in_features=256, out_features=4, bias=True)
            )
          )
          (txt_align_upscaled_embedding): Linear(in_features=768, out_features=96, bias=True)
        )
        (prompt_encoder): PromptEncoder(
          (pe_layer): PositionEmbeddingRandom()
          (point_embeddings): ModuleList(
            (0-3): 4 x Embedding(1, 768)
          )
          (not_a_point_embed): Embedding(1, 768)
          (mask_downscaling): Sequential(
            (0): Conv2d(1, 4, kernel_size=(2, 2), stride=(2, 2))
            (1): LayerNorm2d()
            (2): GELU(approximate='none')
            (3): Conv2d(4, 16, kernel_size=(2, 2), stride=(2, 2))
            (4): LayerNorm2d()
            (5): GELU(approximate='none')
            (6): Conv2d(16, 768, kernel_size=(1, 1), stride=(1, 1))
          )
          (no_mask_embed): Embedding(1, 768)
        )
        (text_encoder): TextEncoder(
          (clip_text_model): CLIPTextModel(
            (text_model): CLIPTextTransformer(
              (embeddings): CLIPTextEmbeddings(
                (token_embedding): Embedding(49408, 512)
                (position_embedding): Embedding(77, 512)
              )
              (encoder): CLIPEncoder(
                (layers): ModuleList(
                  (0-11): 12 x CLIPEncoderLayer(
                    (self_attn): CLIPAttention(
                      (k_proj): Linear(in_features=512, out_features=512, bias=True)
                      (v_proj): lora.Linear(
                        (base_layer): Linear(in_features=512, out_features=512, bias=True)
                        (lora_dropout): ModuleDict(
                          (default): Identity()
                        )
                        (lora_A): ModuleDict(
                          (default): Linear(in_features=512, out_features=8, bias=False)
                        )
                        (lora_B): ModuleDict(
                          (default): Linear(in_features=8, out_features=512, bias=False)
                        )
                        (lora_embedding_A): ParameterDict()
                        (lora_embedding_B): ParameterDict()
                      )
                      (q_proj): lora.Linear(
                        (base_layer): Linear(in_features=512, out_features=512, bias=True)
                        (lora_dropout): ModuleDict(
                          (default): Identity()
                        )
                        (lora_A): ModuleDict(
                          (default): Linear(in_features=512, out_features=8, bias=False)
                        )
                        (lora_B): ModuleDict(
                          (default): Linear(in_features=8, out_features=512, bias=False)
                        )
                        (lora_embedding_A): ParameterDict()
                        (lora_embedding_B): ParameterDict()
                      )
                      (out_proj): Linear(in_features=512, out_features=512, bias=True)
                    )
                    (layer_norm1): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
                    (mlp): CLIPMLP(
                      (activation_fn): QuickGELUActivation()
                      (fc1): Linear(in_features=512, out_features=2048, bias=True)
                      (fc2): Linear(in_features=2048, out_features=512, bias=True)
                    )
                    (layer_norm2): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
                  )
                )
              )
              (final_layer_norm): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
            )
          )
          (dim_align): Linear(in_features=512, out_features=768, bias=True)
        )
        (dice_loss): BinaryDiceLoss()
        (bce_loss): BCELoss(
          (criterion): BCEWithLogitsLoss()
        )
      )
    )
  )
)
```

File: notebooks/train.py
[the remark this answers is not in the document]
Comment (m-krastev) [reply notebooks/train.py:1]: By popular demand, yes

File: adapt_med_seg/models/segvol_lora.py (hunk 1 above)
Comment (m-krastev) [reply adapt_med_seg/models/segvol_lora.py]: Added a couple more.

**Pull request merged**